pkg/s3select/message.go

```diff
@@ -0,0 +1,384 @@
+//
+// Payload specification:
+// End messages have no payload.
+var endMessage = []byte{
+	0, 0, 0, 56, // total byte-length.
+	0, 0, 0, 40, // headers byte-length.
+	193, 198, 132, 212, // prelude crc.
+	13, ':', 'm', 'e', 's', 's', 'a', 'g', 'e', '-', 't', 'y', 'p', 'e', 7, 0, 5, 'e', 'v', 'e', 'n', 't', // headers.
+	11, ':', 'e', 'v', 'e', 'n', 't', '-', 't', 'y', 'p', 'e', 7, 0, 3, 'E', 'n', 'd', // headers.
+	207, 151, 211, 146, // message crc.
+}
+
+// newErrorMessage - creates new Request Level Error Message. S3 sends this message if the request failed for any reason.
+// It contains the error code and error message for the failure. If S3 sends a RequestLevelError message,
+// it doesn't send an End message.
+//
+// Header specification:
+// Request-level error messages contain three headers, as follows:
+// https://docs.aws.amazon.com/AmazonS3/latest/API/images/s3select-frame-diagram-error.png
+//
+// Payload specification:
+// Request-level error messages have no payload.
+func newErrorMessage(errorCode, errorMessage []byte) []byte {
+	buf := new(bytes.Buffer)
+
+	buf.Write([]byte{13, ':', 'm', 'e', 's', 's', 'a', 'g', 'e', '-', 't', 'y', 'p', 'e', 7, 0, 5, 'e', 'r', 'r', 'o', 'r'})
+
+	buf.Write([]byte{14, ':', 'e', 'r', 'r', 'o', 'r', '-', 'm', 'e', 's', 's', 'a', 'g', 'e', 7})
+	binary.Write(buf, binary.BigEndian, uint16(len(errorMessage)))
+	buf.Write(errorMessage)
+
+	buf.Write([]byte{11, ':', 'e', 'r', 'r', 'o', 'r', '-', 'c', 'o', 'd', 'e', 7})
+	binary.Write(buf, binary.BigEndian, uint16(len(errorCode)))
+	buf.Write(errorCode)
+
+	return genMessage(buf.Bytes(), nil)
+}
+
+// NewErrorMessage - creates new Request Level Error Message specified in
+// https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectSELECTContent.html.
+func NewErrorMessage(errorCode, errorMessage string) []byte {
+	return newErrorMessage([]byte(errorCode), []byte(errorMessage))
+}
+
+type messageWriter struct {
+	writer          http.ResponseWriter
+	getProgressFunc func() (int64, int64)
+	bytesReturned   int64
+
+	dataCh  chan []byte
+	doneCh  chan struct{}
+	closeCh chan struct{}
+	stopped uint32
+	closed  uint32
+}
+
+func (writer *messageWriter) write(data []byte) bool {
+	if _, err := writer.writer.Write(data); err != nil {
+		return false
+	}
+
+	writer.writer.(http.Flusher).Flush()
+	return true
+}
+
+func (writer *messageWriter) start() {
+	keepAliveTicker := time.NewTicker(1 * time.Second)
+	var progressTicker *time.Ticker
+	if writer.getProgressFunc != nil {
+		progressTicker = time.NewTicker(1 * time.Minute)
+	}
+
+	go func() {
+		quitFlag := 0
+		for quitFlag == 0 {
+			if progressTicker == nil {
+				select {
+				case data, ok := <-writer.dataCh:
+					if !ok {
+						quitFlag = 1
+						break
+					}
+					if !writer.write(data) {
+						quitFlag = 1
+					}
+				case <-writer.doneCh:
+					quitFlag = 2
+				case <-keepAliveTicker.C:
+					if !writer.write([]byte("{}")) {
```

Comment:
Writing an empty JSON is not supported and it will cause problems if the CSV output format is being used.  Also i think this should be `if !writer.write(continuationMessage) {`

Reply:
done